sdk/core/core/inc/az_action.h

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_ACTION_H
+#define AZ_ACTION_H
+
+#include <az_contract.h>
+
+#include <_az_cfg_prefix.h>
+
+#define AZ_CAT(A, B) A##B
+
+#define AZ_ACTION_ARG(NAME) AZ_CAT(NAME, _arg)
+
+/**
+ * Defines an action @NAME type which accepts an argument of type @ARG.
+ */
+#define AZ_ACTION_TYPE(NAME, ARG) \
+  typedef ARG AZ_ACTION_ARG(NAME); \
+  typedef struct { \
+    az_result (*func)(void *, ARG); \
+    void *self; \
+  } NAME; \
+  AZ_NODISCARD AZ_INLINE az_result AZ_CAT(NAME, _do)(NAME const action, ARG const arg) { \
+    AZ_CONTRACT_ARG_NOT_NULL(action.func); \
+    return action.func(action.self, arg); \
+  }
+
+/**
+ * Defines a function @NAME##_action which creates an action of type @ACTION
+ * using the given @NAME function.
+ *
+ * The function is a safe way to create an action with the same type @SELF for 
+ * the first argument of the function @NAME and @ACTION.self.
+ */
+#define AZ_ACTION_FUNC(NAME, SELF, ACTION) \
```

Comment:
ditto about the macros.

It's really, really, really hard to figure out what code is actually doing when it starts to use and combine macros like this.

Reply:
I put some notes about principals how to work with actions here https://github.com/Azure/azure-sdk-for-c/blob/sergey/refactoring/emitter-design.md It's work in progress. Anyway, we already had this macro in `az_callback.h`